File: e2e/java/src/test/java/dev/kreuzberg/e2e/RateLimitTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/** E2e tests for category: rate_limit. */
class RateLimitTest {
    @Test
    void testRateLimitBasicDelay() throws Exception {
        // Rate limiter adds delay between requests to the same domain
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(3, result.crawl().pagesCrawled());
        assertTrue(result.rateLimit().minDurationMs() >= 150, "expected >= 150");
    }

    @Test
    void testRateLimitZeroNoDelay() throws Exception {
        // Rate limiter with zero delay does not slow crawling
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(2, result.crawl().pagesCrawled());
    }

}

File: e2e/php/tests/RateLimitTest.php
<?php

declare(strict_types=1);

namespace Kreuzberg\E2e;

use PHPUnit\Framework\TestCase;
use Kreuzcrawl\Kreuzcrawl;

/** E2e tests for category: rate_limit. */
final class RateLimitTest extends TestCase
{
    /** Rate limiter adds delay between requests to the same domain */
    public function test_rate_limit_basic_delay(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(3, $result->crawl->pages_crawled);
        $this->assertGreaterThanOrEqual(150, $result->rate_limit->min_duration_ms);
    }

    /** Rate limiter with zero delay does not slow crawling */
    public function test_rate_limit_zero_no_delay(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(2, $result->crawl->pages_crawled);
    }
}
